### DISCUSSION TITLE
超量使用 API 錯誤是什麼？

### DISCUSSION THREAD

**Discussion by osk2** [osk2/panasonic_smart_app#31] [Announcements]
## 說明

使用整合時可能會在 log 中看見「超量使用 API，功能將受限制」的訊息，這表示你的帳號目前被伺服器封鎖，
這時除 Home Assistant 整合外，App （v2.0.0之前的版本）也會無法使用

被封鎖時套件會自動切換至備用 API，但仍會有部分功能無法使用：

#### 冷氣

冷氣控制功能可正常使用，但無法更新狀態，除下列外：

- 開關狀態
- 目前設定的溫度

#### 除濕機

除濕機控制功能可正常使用，但無法更新狀態，除下列外：

- 開關狀態
- 目前設定的濕度

待解除封鎖後即會自動恢復完整功能

## 解決方法

📣 更新：v2.6.0 開始支援 proxy，可用 proxy 定期更換 IP 的方式來避開封鎖（[使用說明](https://github.com/osk2/panasonic_smart_app/discussions/81)）

因超量使用被伺服器封鎖通常會在 12~24 小時內解除封鎖，可試著在設定整合時調高更新頻率（預設為每 180 秒更新一次）

整合套件在無法得知封鎖條件及方法的情況下沒有辦法避免被封鎖，請向原廠反應此問題，要求公佈 API 封鎖的原則

## 已知問題

使用備用 API 時，裝置狀態可能會變成相反的（沒錯，就是這麼雷⋯），若遇到此問題請向原廠反應

## 備註

Panasonic Taiwan 聯絡信箱：voc.care@tw.panasonic.com

**Comment by toddchou** [comment]
我買了三台除濕機,以下是Panasonic 台灣的回覆,有這種限制真的很low, 本來要買第四台Panasonic 聯網除濕機,現在已經不考慮了
----------------------------------------------------------
親愛的 Panasonic 愛用者 先生/小姐：您好，

我是台灣松下顧客商談中心；敝姓X，

非常感謝您對Panasonic商品的愛用與支持。

 

一個帳號最多可以在六個設備上登入，

但如果在同一個網域內頻繁操作時，也有可能無法讀取。

建議您先停止使用第三方APP或程式登入或控制除濕機，

以避免此情形再次發生。

-----------------------------

**Reply by osk2** [reply]
感謝情報，我也曾建議 Panasonic 乾脆開放內網控制，但客服只是官腔回覆真的是很讓人失望

**Reply by toddchou** [reply]
以下是最不能理解的,還限制使用者的使用習慣,Panasonic TW 在這各領域,應該還有很多東西要用心經營
"但如果在同一個網域內頻繁操作時，也有可能無法讀取。"

**Reply by toddchou** [reply]
> 感謝情報，我也曾建議 Panasonic 乾脆開放內網控制，但客服只是官腔回覆真的是很讓人失望

這才是好方法, 這家pansonic TW 真的是台式作風, 要認真學學phlips hue, smatthing ,Ring

**Reply by toddchou** [reply]
目前polling interview 應經改為480
還是每天都會遇到這個問題, 
想說我買了三台除濕機
就給他增加預設值
每天都會遇到這個問題

**Reply by osk2** [reply]
最近有在思考要改一下作法，把每個裝置的更新間隔拉長，應該就能避免掉這個問題，但代價就是更新過程會很耗時，實在是兩難 😓

**Reply by Adios** [reply]
會不會有一種可能的方式：直接連 IoT devices？

HACS 兩個 TP-Link 的 integrations: tapo camera, tapo plug,

它們似乎都是直接打機器（這兩個產品正常使用下，都是透過 app 連 tp 的 cloud 再控制）

我還沒看他們的做法 (也許 tapo device 本身有 undocumented API 之類的)

先提出個想法看看

**Reply by osk2** [reply]
我自己也有 Tapo 攝影機，它本身就留有本地控制的介面
之前試過掃國際牌模組開的 port，看起來是沒有走本地控制的機會

目前可以用2.6版搭配會自動更換 IP 的 proxy 就能避開封鎖機制了

**Reply by Adios** [reply]
啊真是令人難過
目前好像也只能先換 ip 來避 qq

**Reply by morrisjazz** [reply]
請問我使用Nord VPN的Proxy但是無法連線成功
請問有什麼推薦的嗎？

**Reply by osk2** [reply]
我也不確定哪些可以正常連線，我自己是沒有用 proxy，是 @OTZro 在[這裡](https://github.com/osk2/panasonic_smart_app/issues/72)分享的

**Reply by OTZro** [reply]
我目前用的是IPRoyal, 這個 proxy 可以定期換ip
這類型的proxy滿多的,可以搜尋rotating proxy並比較看看

**Reply by OTZro** [reply]
我有想過另一種方式是利用 dnsspoof 的方式來做 man-in-the-middle attack，回報機器狀態到 panasonic server 的時候同時也把資料同步送到 mqtt，這樣做的好處是可以很即時的知道機器的狀態，並且不會踩到 panasonic api 超量的問題。不過實作起來滿花時間的，並且也不知道實際溝通的協定，我這邊就沒有去研究。
這個部份如果有大大有興趣可以嘗試看看。

**Reply by osk2** [reply]
其實以前有試過開發中間人攻擊，印象中早期韌體的連線是沒加密的，後期的版本就加上去了，後來忙起來就沒有再深入研究這部份了
不然我認為是蠻好的作法，就是要實作的複雜度真是蠻高的

**Reply by Adios** [reply]
今天有 firmware update 就好奇 mitmproxy 會看到什麼東西，更新過程的截圖：

<img width="1004" alt="スクリーンショット 2023-07-19 午後4 42 11" src="https://github.com/osk2/panasonic_smart_app/assets/62105/2ede8a57-b378-40cd-adb9-6df0b39cc50a">

- firmware 更新時候會放棄 wss 用傳統的 http，app 會下載 firmware 至手機中
  - 再 `POST http://DEHUMIDIFIER_IP:57223/Upgrade` 叫除濕機 TFTP 過來抓抓
- `http://DEHUMIDIFIER_IP:57223/device.xml` (and `SmartHomeSCPD.xml`)
  - full xml: https://gist.github.com/Adios/07c0b9efa143d0b1fca766155f3b17b6
    - 有關鍵字 `<controlURL>/SmartHome/Control</controlURL>`
    - 但似乎是 firmware 自身 (SAANet mod) 維護操作...嗎?
- firmware: `http://ems.panasonic.com.tw/PSHEFileUpload/CZ-T006/tw-214_T006-ota.bin`
  - `strings tw-214_T006-ota.bin`
    - 能看到大概的科技
      - `wss://miproxy.panasonic.com.tw` 和訊息格式
    - 有時間能再看看有沒有什麼 hidden path or string..?
    - e.g.: ***I@Love@Panasonic@Smart*** XD

----
- port 也許不一定是 57223? 搞不好是 app 收到更新叫 hosts 開隨機 port
  - strings firmware 沒看到 57223，啊不過有可能是直接編成 int
- 另一塊還可以再 mitm 看看的會是 device 剛註冊直連 PanasonicSmartWifi(?) 的時候.. 之後
- 今天有嘗試 spoofing `miproxy.panasonic.com.tw`
  - 但當時我沒有意識到是 wss
  - device 一直連我噴 error 最後大家都連不上 device 要重設 device wifi 才行 XDD
- `192.168.0.161`: F-Y20JH
- `192.168.0.156`: LJ36 + CZ-T007，兩台都是用同一個 firmware 更新

**Reply by osk2** [reply]
我開了一個新的[討論串](https://github.com/osk2/panasonic_smart_app/discussions/77)，各位有興趣的可以移駕到那討論 👏

**Comment by che-chun** [comment]
如果先以有實際操作(如更改開關/調溫度..等)時，在進行裝置狀態的更新，會不會有效減少對伺服器的請求？
當然如果本身有用到偵測溫度就做什麼的..自動化設定，可能就會失效了。

**Comment by Adios** [comment]
陸續入手三台除機姬想說可以串 HA 結果半天不到就被 log 導來這裡了...

有趣的是現在超量的當下，只有 iOS 的 panasonic smart app 顯示不出資訊，Android 三台都正常顯示

**Comment by yourcst** [comment]
真的不建議買Panasonic的智慧商品,我去年搬新家就是看上他們擁有所謂的智慧型商品才買了六台冷氣,其中5台為的整合到HA,考慮減少問題所以直接就上了項規,雖然成功整合進HA但馬上就出現了超量使用 API問題.我寫了Email通報了相關整合上出現的問題.但他們似乎不想回答這種問題.最後只能撥打電話過濾掉API問題後只通報屬於5台冷氣硬體上出現的問題(因設計不良互相干擾而自動關機問題,其中主臥的冷氣不知什麼原因總是會產生很大的噪音,技師來家調整了很多次到最後我只能試著去習慣隨時會出現的聲音).他們才會很樂意的幫我處理.如果他們不能處理掉API的問題.我應該不會再去買他家的產品了.因為真的讓我在使用HA上很是無言.
